osu.Game/Utils/ModUtils.cs

```diff
@@ -112,14 +112,34 @@ public static bool CheckAllowed(IEnumerable<Mod> combination, IEnumerable<Type>
         /// <param name="invalidMods">Invalid mods, if any were found. Can be null if all mods were valid.</param>
         /// <returns>Whether the input mods were all valid. If false, <paramref name="invalidMods"/> will contain all invalid entries.</returns>
         public static bool CheckValidForGameplay(IEnumerable<Mod> mods, [NotNullWhen(false)] out List<Mod>? invalidMods)
+            => checkValid(mods, m => m.Type != ModType.System && m.HasImplementation && !(m is MultiMod), out invalidMods);
+
+        /// <summary>
+        /// Check the provided combination of mods are valid for a multiplayer match session.
+        /// </summary>
+        /// <param name="mods">The mods to check.</param>
+        /// <param name="invalidMods">Invalid mods, if any were found. Can be null if all mods were valid.</param>
+        /// <returns>Whether the input mods were all valid. If false, <paramref name="invalidMods"/> will contain all invalid entries.</returns>
+        public static bool CheckValidForMultiplayer(IEnumerable<Mod> mods, [NotNullWhen(false)] out List<Mod>? invalidMods)
+            => checkValid(mods, m => m.PlayableInMultiplayer, out invalidMods);
+
+        /// <summary>
+        /// Check the provided combination of mods are valid as "free mods" in a multiplayer match session.
+        /// </summary>
+        /// <param name="mods">The mods to check.</param>
+        /// <param name="invalidMods">Invalid mods, if any were found. Can be null if all mods were valid.</param>
+        /// <returns>Whether the input mods were all valid. If false, <paramref name="invalidMods"/> will contain all invalid entries.</returns>
+        public static bool CheckValidFreeModsForMultiplayer(IEnumerable<Mod> mods, [NotNullWhen(false)] out List<Mod>? invalidMods)
+            => checkValid(mods, m => m.ValidFreeModInMultiplayer, out invalidMods);
+
+        private static bool checkValid(IEnumerable<Mod> mods, Predicate<Mod> valid, [NotNullWhen(false)] out List<Mod>? invalidMods)
         {
             mods = mods.ToArray();
-
-            CheckCompatibleSet(mods, out invalidMods);
```

Comment:
I'll probably have to mention it in xmldoc, but with `CheckCompatibleSet`, this can potentially lead to checking compatibility more than once at the same time (see ppy/osu-server-spectator#106).

Reply:
Not super convinced by any of the above. Checking compatibility twice can be avoided server-side by just removing the `CheckCompatibleSet()` call and only leaving the `CheckValidForMultiplayer()` check. `CheckValidFreeModsForMultiplayer()` shouldn't be checking compatibility anyways.

Going to hold out for more opinions on this.

Reply:
> Not super convinced by any of the above. Checking compatibility twice can be avoided server-side by just removing the CheckCompatibleSet() call and only leaving the CheckValidForMultiplayer() chec

Hard to make that visible by combining them in one exception message, that's exactly the reason I've done that. 

> CheckValidFreeModsForMultiplayer() shouldn't be checking compatibility anyways.

With the change you're proposing, compatibility will be checked in "free mods" context as well.

Reply:
> With the change you're proposing, compatibility will be checked in "free mods" context as well.

Except you don't have to put that `CheckCompatibleSet()` call back into `checkValid()`? That would just be incorrect. The call could however be added to `CheckValidForGameplay()` and `CheckValidForMultiplayer()` one level higher.